impl/src/test/java/org/ehcache/impl/internal/store/offheap/AssertingLazyOffHeapValueHolder.java

```diff
@@ -0,0 +1,104 @@
+package org.ehcache.impl.internal.store.offheap;
+
+import org.ehcache.spi.serialization.Serializer;
+import org.terracotta.offheapstore.storage.portability.WriteContext;
+
+import java.nio.ByteBuffer;
+import java.util.Collection;
+import java.util.function.Predicate;
+
+import static java.util.Arrays.asList;
+import static java.util.Arrays.stream;
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertThat;
+
+public class AssertingLazyOffHeapValueHolder<V> extends LazyOffHeapValueHolder<V> {
+
+  /**
+   * This is a set of 'patterns' that capture a subset of the lock scopes that the `LazyOffHeapValueHolder` can be
+   * called from.  You might end up looking at these patterns for one of two reasons:
+   *
+   * 1. (Most Likely) You edited one of the files here and moved a lock scope to a new line number range.  In this case
+   *    you can just update the line number range (inclusive bounds at both ends) and carry on your way.
+   * 2. (Quite Likely) You introduced new code or new testing that access a pre-existing lock scope that is not listed
+   *    here.  Find the lock scope in the stack trace of the thrown exception and add an appropriate stanza here.
+   * 2. (Less Likely) You introduced a new call path that leaks a still serialized, attached value holder that someone
+   *    else tried to access later (outside lock scope).  In this case you must locate the source of the value holder
+   *    and either force deserialization or detach the value under the pre-existing lock scope.
+   * 3. (Least Likely) You introduced a new method to OffHeapStore that needs locking properly.
+   */
+  private static final Collection<Predicate<StackTraceElement>> LOCK_SCOPES = asList(
+    className("org.terracotta.offheapstore.AbstractLockedOffHeapHashMap").methodName("shrink").lineRange(499, 501),
+    className("org.terracotta.offheapstore.AbstractLockedOffHeapHashMap").methodName("computeWithMetadata").lineRange(519, 521),
+    className("org.terracotta.offheapstore.AbstractLockedOffHeapHashMap").methodName("computeIfPresentWithMetadata").lineRange(541, 543),
+    className("org.ehcache.impl.internal.store.offheap.factories.EhcacheSegmentFactory$EhcacheSegment$EntrySet").methodName("iterator").lineRange(125, 135),
+    className("org.ehcache.impl.internal.store.disk.factories.EhcachePersistentSegmentFactory$EhcachePersistentSegment$EntrySet").methodName("iterator").lineRange(129, 138),
+    className("org.terracotta.offheapstore.AbstractLockedOffHeapHashMap$LockedEntryIterator").methodName("next").lineRange(391, 393)
```

Comment:
How about reading the bytecode to discover the line ranges? With Javassist:

```
  static class Range {
    int begin = -1;
    int end = -1;
  }

  private static Range findLockRange(String className, String methodName) throws BadBytecode, NotFoundException {
    ClassPool pool = ClassPool.getDefault();
    CtClass ctClass = pool.get(className);
    CtMethod ctMethod = ctClass.getDeclaredMethod(methodName);
    MethodInfo methodInfo = ctMethod.getMethodInfo2();
    CodeAttribute codeAttribute = methodInfo.getCodeAttribute();
    ConstPool constPool = methodInfo.getConstPool();
    CodeIterator codeIterator = codeAttribute.iterator();

    Range range = new Range();
    while (codeIterator.hasNext()) {
      int pos = codeIterator.next();

      if (range.begin == -1) {
        range.begin = lineNumberOfInvokeInterface(codeIterator, pos, constPool, methodInfo,"java.util.concurrent.locks.Lock","lock");
        if (range.begin != -1) {
          range.begin++;
        }
      } else if (range.end == -1) {
        range.end = lineNumberOfInvokeInterface(codeIterator, pos, constPool, methodInfo,"java.util.concurrent.locks.Lock","unlock");
        if (range.end != -1) {
          range.end--;
        }
      }

      if (range.end != -1) {
        return range;
      }
    }
    throw new AssertionError(ctMethod.getName() + " does not contain both calls to Lock.lock and Lock.unlock");
  }

  private static int lineNumberOfInvokeInterface(CodeIterator iter, int pos, ConstPool pool, MethodInfo methodInfo, String intfName, String methodName) {
    int opcode = iter.byteAt(pos);
    if (opcode == INVOKEINTERFACE) {
      int index = iter.u16bitAt(pos + 1);
      String interfaceMethodrefClassName = pool.getInterfaceMethodrefClassName(index);
      if (!interfaceMethodrefClassName.equals(intfName)) {
        return -1;
      }
      String interfaceMethodrefName = pool.getInterfaceMethodrefName(index);
      if (!interfaceMethodrefName.equals(methodName)) {
        return -1;
      }
      return methodInfo.getLineNumber(pos);
    }
    return -1;
  }

```

Reply:
It would make the test less brittle at the potential cost of an additional test dependency.  Not a bad trade-off.

Reply:
Will look in to this.

Reply:
The ASM code looks good to me